pkg/event-consumers/nats/nats-consumer.go

```diff
@@ -0,0 +1,178 @@
+
+	nc, err := nats.Connect(url)
+	if err != nil {
+		logrus.Fatalf("Failed to initialize nats client to server %v due to %v", url, err)
+	}
+	defer nc.Close()
+
+	logrus.Infof("Joining to the queue group %v and subscribing to the topic %v", queueGroupID, topic)
+	subscription, err := nc.QueueSubscribe(topic, queueGroupID, func(msg *nats.Msg) {
+		logrus.Debugf("Received Message %v on Topic: %v Queue: %v", string(msg.Data), msg.Subject, msg.Sub.Queue)
+		logrus.Infof("Sending message %s to function %s", string(msg.Data), funcName)
+		req, err := getHTTPReq(clientset, funcName, ns, "POST", string(msg.Data))
+		if err != nil {
+			logrus.Errorf("Unable to elaborate request: %v", err)
+		} else {
+			//forward msg to function
+			err = sendMessage(req)
+			if err != nil {
+				logrus.Errorf("Failed to send message to function: %v", err)
+			} else {
+				logrus.Infof("Message has sent to function %s successfully", funcName)
+			}
+		}
+	})
+	if err != nil {
+		logrus.Fatalf("Failed to create queue group %v and subscribing to the topic %v", queueGroupID, topic)
+	}
+	defer close(stoppedchan)
+	for {
+		select {
+		case <-stopchan:
+			err = subscription.Unsubscribe()
+			if err != nil {
+				logrus.Fatalf("Failed to unsubscribe from the queue group %v and topic %v", queueGroupID, topic)
+			}
+			return
+		}
+	}
+}
+
+func isJSON(s string) bool {
+	var js map[string]interface{}
+	return json.Unmarshal([]byte(s), &js) == nil
+
+}
+
+func getHTTPReq(clientset kubernetes.Interface, funcName, namespace, method, body string) (*http.Request, error) {
+	svc, err := clientset.CoreV1().Services(namespace).Get(funcName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("Unable to find the service for function %s", funcName)
+	}
+	funcPort := strconv.Itoa(int(svc.Spec.Ports[0].Port))
+	req, err := http.NewRequest(method, fmt.Sprintf("http://%s.%s.svc.cluster.local:%s", funcName, namespace, funcPort), strings.NewReader(body))
+	if err != nil {
+		return nil, fmt.Errorf("Unable to create request %v", err)
+	}
+	timestamp := time.Now().UTC()
+	eventID, err := utils.GetRandString(11)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create a event-ID %v", err)
+	}
+	req.Header.Add("event-id", eventID)
+	req.Header.Add("event-time", timestamp.String())
+	req.Header.Add("event-namespace", "natstriggers.kubeless.io")
+	if isJSON(body) {
+		logrus.Infof("TRUE %v", body)
+		req.Header.Add("Content-Type", "application/json")
+		req.Header.Add("event-type", "application/json")
+	} else {
+		logrus.Infof("FLASE %v", body)
+		req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+		req.Header.Add("event-type", "application/x-www-form-urlencoded")
+	}
+	return req, nil
+}
+
+func sendMessage(req *http.Request) error {
+	client := &http.Client{}
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+	if res.StatusCode != 200 {
```

Comment:
should do something like `defer resp.Body.Close()` here to close response body and avoid leaking file descriptors ;)